File: flow.cc
#include "flow.h"

#include <boost/functional/hash.hpp>
#include <sstream>

Flow::Flow() {
	this->src = 0;
	this->dst = 0;
	this->srcPort = 0;
	this->dstPort = 0;
	this->type = FlowType::TCP;
}

Flow::Flow(uint32_t src, int srcPort, uint32_t dst, int dstPort, FlowType::Type flowType) {
	this->src = src;
	this->srcPort = srcPort;
	this->dst = dst;
	this->dstPort = dstPort;
	this->type = flowType;
}

const uint32_t Flow::getDst() const {
	return dst;
}

void Flow::setDst(const uint32_t dst) {
	this->dst = dst;
}

int Flow::getDstPort() const {
	return dstPort;
}

void Flow::setDstPort(int dstPort) {
	this->dstPort = dstPort;
}

const uint32_t Flow::getSrc() const {
	return src;
}

void Flow::setSrc(const uint32_t src) {
	this->src = src;
}

int Flow::getSrcPort() const {
	return srcPort;
}

void Flow::setSrcPort(int srcPort) {
	this->srcPort = srcPort;
}

FlowType::Type Flow::getType() const {
	return type;
}

std::string Flow::getTypeStr() const {
	return flowTypeStr(type);
}

void Flow::setType(FlowType::Type type) {
	this->type = type;
}

const std::string Flow::toString() const {
	std::stringstream sstm2;
	sstm2 << src << ":" << srcPort
			<< " --> "
			<< dst << ":" << dstPort
			<< ", " << type;

	return sstm2.str();
}

const std::string Flow::toFormattedString() const {
	std::stringstream sstm2;
	sstm2 << std::fixed;
	sstm2 << src << ":" << srcPort << "\t"
			<< dst << ":" << dstPort << "\t"
			<< type;

	return sstm2.str();
}

Flow::~Flow() {
}

bool Flow::operator ==(const Flow& ref) const {
	if(this->src == ref.getSrc()
			&& this->srcPort == ref.getSrcPort()
			&& this->dst == ref.getDst()
			&& this->dstPort == ref.getDstPort()
			&& this->type == ref.getType()){
		return true;
	} else {
		return false;
	}
}

bool Flow::operator <(const Flow& ref) const {
	return ((this->src + this->srcPort + this->dst + this->dstPort) < (ref.getSrc() + ref.getSrcPort() + ref.getDst() + ref.getDstPort()));
}

int Flow::hashCode() {
	boost::hash<std::string> string_hash;

	return string_hash(toString());
}

File: flow.py
import re

class Flow:
    def __init__(self, src='', sport=0, dst='', dport=0, flowType=0):
        self.src = src
        self.sport = int(sport)
        self.dst = dst
        self.dport = int(dport)
        if(flowType == 'UDP'):
            self.flowType == 0
        elif(flowType == 'TCP'):
            self.flowType == 1
        else:
            self.flowType = int(flowType)
    def __eq__(self, flow):
        if(self.src == flow.src and
              self.sport == flow.sport and
              self.dst == flow.dst and
              self.dport == flow.dport and
              self.flowType == flow.flowType):
            return True
        else:
            return False
    def __ne__(self, flow):
        return not self.__eq__(flow)
    def parse(self, string):
        items = re.split(r'[\s :]', string)
        self.src = items[0]
        self.sport = int(items[1])
        self.dst = items[2]
        self.dport = int(items[3])
        if(items[4] == 'UDP'):
            self.flowType = 0
        elif(items[4] == 'TCP'):
            self.flowType = 1
        else:
            self.flowType = int(items[4])

    def to_string(self):
        out = 'flow = {}:{} {}:{} {}'.format(self.src, self.sport, self.dst, self.dport, self.flowType)
        return out
#
# Testcode
#
#f1 = Flow('10.1.1.1', 6677, '10.1.1.4', 5004, 0)
#f2 = Flow('10.1.1.10', 6677, '10.1.1.5', 5004, 0)
#f3 = Flow('10.1.1.1', 6677, '10.1.1.4', 5004, 0)
#print f1 == f2
#print f1 == f3
#f3.parse('10.1.1.5:6655 10.1.1.7:443 1')
#print f3.dst
#f4 = Flow()
#print f4.sport
#f4.parse('10.1.1.5:6655 10.1.1.7:443 1')
#print f4.sport
